hadoop-hdds/docs/content/design/container-reconciliation.md

```diff
@@ -0,0 +1,246 @@
+    - Instead, schedule the imported container for on-demand scanning to be sure SCM knows of corruption if it does not already.
+    - **Optimization**: Clear last scanned timestamp on container import to make the background scanner get to it faster if datanode restarts before the on-demand scan completes.
+
+#### On `getContainerHashes`
+
+- Return the current Merkle Tree of the specified container.
+    - The tree will be calculated from existing chunk checksums if it is not present. However, this should not happen in the majority of cases since it will be calculated on container close.
+- If container not closed, return error.
+- If container not found, return error.
+
+- The merkle tree returned by this call will always be consistent.
+    - The hash of each node in the tree will always accurately represent a hash of its children.
+- The merkle tree returned by this call may not always match the current state of the container.
+    - The container can be repaired or corrupted any time between the scan that generated the tree and the call to read the container's tree.
+    - If the caller decides it needs to read a chunk based on the merkle tree obtained, it will have to re-check the checksum obtained as part of the chunk read to verify the data received matches what it was expecting.
+
+#### On `reconcileContainer`
+
+- Use existing merkle tree, unless its not there, then compute it.
+    - The tree will be calculated from existing chunk checksums if it is not present. However, this should not happen in the majority of cases since it will be calculated on container close.
+1. For each replica:
+    1. Datanode calls `getContainerHashes`
+    2. Datanode diffs this with its current merkle tree to find all chunks that are:
+        1. Not part of blocks that are marked as deleted in this or the other replica's tree.
+        2. Not matching the chunk's checksum value in this datanode's tree.
+    3. For each chunk in the other replica's tree that does not match this replica's tree:
+        1. If the chunk or block is not present in our container, save it in memory as needing to be imported.
+        2. If the chunk or block is not marked `healthy` in our merkle tree (known to be corrupted by the scanner), save it in memory as needing to be reconciled.
+        3. If chunk checksums differ but our chunk is marked `healthy`, ignore this chunk.
+            - If the chunk is corrupted in the other replica or got corrupted while we were reading it, we should not import it.
+            - If the chunk is corrupted on one of the replicas but the `healthy` flag has not been updated yet to reflect this, it will be resolved when the scanner reaches the container again.
+            - If the same chunk is truly `healthy` in both replicas but the checksums do not match, there has been an irrecoverable error on the write path where different data was written to each replica but they both passed checksum validation.
```

Comment:
We probably need a way to flag such a block and exclude it from the container hash. Otherwise the replica hashes will always different and the container will always been seen as "needs reconciliation" from SCMs perspective.

Reply:
Good point.

Reply:
SCM co-ordination is not in scope for this document for now.